### ISSUE TITLE
Bad PrometheusRule prevented prometheus operator from starting

### ISSUE THREAD

**Issue by corlettb** [coreos/prometheus-operator#2727]
The prometheus operator was failing to start with the following error:

```
msg="Unhandled error received. Exiting..." 
err="creating CRDs failed: waiting for PrometheusRule crd failed: timed out waiting for Custom Resource: failed to list CRD: v1.PrometheusRuleList.Items: []*v1.PrometheusRule: v1.PrometheusRule.Spec: v1.PrometheusRuleSpec.Groups: []v1.RuleGroup: v1.RuleGroup.Rules: []v1.Rule: v1.Rule.Annotations: ReadString: expects \" or n, but found f, error found in #10 byte of ...|summary\":false},\"exp|..., bigger context ...|un successfully in the last 24 hours.\",\"summary\":false},\"expr\":\"job:last_success:hours_ago{ exported|..."
```

This turned out to be a problem with the following rule:

```
---
apiVersion: monitoring.coreos.com/v1
kind: PrometheusRule
...
spec:
  groups:
    - name: fakerule.rules
      rules:
        - alert: SomeAlert
          expr: someexpression > 0
          for: 1m
          labels:
            severity: critical
          annotations:
            summary: No
            description: some alert
```

The problem part of this is the "summary: No". No is special in yaml and will be resolved as a boolean that is false.

Now granted this is bad config, but the operator should deal with this better.

- The operator should log the namespace and path for the invalid crd in the error message.
- A bad prometheusrule should not prevent the operator from starting. Invalid rules should just be skipped over.
- The operator should update a "status" field on the crd to inform the user that their prometheusrule has been rejected and why.
- The operator should expose a metric to prometheus to show failing rules.

**Comment by allen13** [comment]
We have been successfully using the built-in admission webhook in production for the past month now. See implementation details here: https://github.com/helm/charts/tree/master/stable/prometheus-operator/templates/prometheus-operator/admission-webhooks

**Comment by caarlos0** [comment]
similar issue here, but maybe a bit worse:

if the operator is already running and the config is updated to a "bad" one, it keeps running, but throwing that error every time and basically not working.

also no metrics are exposed about this, I expected that `prometheus_operator_rule_validation_errors_total` would increase - or maybe another metrics.

IMHO, increasing that counter would be a good way to inform that things are not working, as I just leave our prometheus broken for a few days without realizing it :( 

if you think this is a good idea and can point me more or less to where I should look at, I can work on a PR 👍 

thanks for this great project BTW 🌟

**Comment by simonpasquier** [comment]
@caarlos0 can you share logs?

> also no metrics are exposed about this, I expected that prometheus_operator_rule_validation_errors_total would increase - or maybe another metrics.

`prometheus_operator_rule_validation_errors_total` are only incremented if you have the admission webhooks configured. Is it your case?

> IMHO, increasing that counter would be a good way to inform that things are not working, as I just leave our prometheus broken for a few days without realizing it :(

You probably want a [`Watchdog`](https://github.com/coreos/kube-prometheus/blob/b8f513e4d485db099a1dec4dffd219518b4d8d57/manifests/prometheus-rules.yaml#L1155-L1165) alert + DeadMansSnitch-like integration to ensure that your alerting pipeline is always functional.

**Comment by vsliouniaev** [comment]
The problem here is that prometheus operator is unable to de-serialize the resource because it wants `string` but gets `bool`. I believe you will have `prometheus_operator_reconcile_errors_total` increase in this case, since it's an issue there.

Admission hooks won't care about this because the deserialization logic there uses prometheus, rather than prometheus-operator code to perform unmarshalling and this passes code according to that spec.

The watchdog alert will keep firing if it was firing previously, the only difference will be in that updates to prometheus will stop being applied until the bad rule is removed.

**Comment by caarlos0** [comment]
@simonpasquier:

> @caarlos0 can you share logs?

is the same as OP, but was for a `true` instead of `'true'`:

```
thanos-prometheus-operator-operator-c7d84cbc6-sbvk6 prometheus-operator E1126 14:34:58.648908       1 reflector.go:125] github.com/coreos/prometheus-operator/pkg/prometheus/operator.go:451: Failed to list *v1.PrometheusRule: v1.PrometheusRuleList.Items: []*v1.PrometheusRule: v1.PrometheusRule.Spec: v1.PrometheusRuleSpec.Groups: []v1.RuleGroup: v1.RuleGroup.Rules: []v1.Rule: v1.Rule.Labels: ReadString: expects " or n, but found t, error found in #10 byte of ...|:{"page":true,"sever|..., bigger context ...|olor=\"red\"} == 1","for":"10s","labels":{"page":true,"severity":"critical"}},{"alert":"ElasticSearc|...
```

> `prometheus_operator_rule_validation_errors_total` are only incremented if you have the admission webhooks configured. Is it your case?

it is not, I got some other issue with that and never fixed it, will take a look, thanks!


> You probably want a [`Watchdog`](https://github.com/coreos/kube-prometheus/blob/b8f513e4d485db099a1dec4dffd219518b4d8d57/manifests/prometheus-rules.yaml#L1155-L1165) alert + DeadMansSnitch-like integration to ensure that your alerting pipeline is always functional.

we have, the thing is just that alert rules were not updating, so we had some old alert rules there...  as @vsliouniaev mention

---

@vsliouniaev:

> The problem here is that prometheus operator is unable to de-serialize the resource because it wants `string` but gets `bool`. I believe you will have `prometheus_operator_reconcile_errors_total` increase in this case, since it's an issue there.

it didn't increase here...

**Comment by allen13** [comment]
Are there any plans on getting the invalid label issue fixed in the validator as mentioned in #2962 ? With a large number of users this breaks the operator on a fairly regular basis.

**Comment by allen13** [comment]
Added PR #3032 to fix the above issues.
